src/hotspot/share/runtime/synchronizer.cpp

```diff
@@ -1291,13 +1309,30 @@ void ObjectSynchronizer::inflate_helper(oop obj) {
 
 // Can be called from non JavaThreads (e.g., VMThread) for FastHashCode
 // calculations as part of JVM/TI tagging.
-static bool is_lock_owned(Thread* thread, oop obj) {
+static bool is_lock_owned(JavaThread* locking_thread, oop obj) {
   assert(LockingMode == LM_LIGHTWEIGHT, "only call this with new lightweight locking enabled");
-  return thread->is_Java_thread() ? JavaThread::cast(thread)->lock_stack().contains(obj) : false;
+  return locking_thread != nullptr ? locking_thread->lock_stack().contains(obj) : false;
 }
 
 ObjectMonitor* ObjectSynchronizer::inflate(Thread* current, oop object,
                                            const InflateCause cause) {
+  JavaThread* locking_thread = nullptr;
+  if (LockingMode == LM_LIGHTWEIGHT && current->is_Java_thread()) {
+    locking_thread = JavaThread::cast(current);
+  }
+  return inflate(locking_thread, current, object, cause);
```

Comment:
This looks strange passing locking_thread as nullptr.  Why not unconditionally make it current?  How can it ever be null?

Reply:
hm, LEGACY doesn't use locking_thread.

Reply:
I agree this doesn't quite makes sense to me. `locking_thread` is really `presumed_owner_thread` - we aren't trying to lock here, we are inflating an already locked object, where the passed in thread is the owner.

Reply:
I very much agree. This stems from the same issue that cause me to type

> `inflate` care about a `locking_thread` is a little unpleasant in my opinion

Inflate in all locking modes has nothing to do with what is inflating. It is just that in LM_LIGHTWEIGHT it is also given the responsibility to fix the owner and the lock stack if the `current` thread is the owner. Because the idea is that modifying the lock stack and moving the owner field from anonymous is only done from the owning thread. This is true everywhere except for re-lock, which is the only place that must enter on behalf of another thread. Which means that it must also inflate, but it must set the correct owner, so it must inflate as if it is inflated from another thread.

`presumed_owner_thread` is not really correct either.

`Inflate` just means create an `ObjectMonitor` if it does not exist and return the `ObjectMonitor*` associated with this object. Regardless if it is locked or not. 

I like the idea of having `inflate_for`, I think the name can be clearer  this is an API that is exclusively needed for LM_LIGHTWEIGHT.